pallets/phala-world/src/nft_sale.rs

```diff
@@ -0,0 +1,1686 @@
+			// Change status of preorder and add to PreorderResult
+			let mut preorder_info =
+				Preorders::<T>::take(preorder_id).ok_or(Error::<T>::NoAvailablePreorderId)?;
+			preorder_info.preorder_status = status.clone();
+			PreorderResults::<T>::insert(&preorder_info.owner, preorder_id, &preorder_info);
+
+			Self::deposit_event(Event::PreorderResultChanged { preorder_id, status });
+
+			Ok(())
+		}
+
+		/// Claim function for bulk minting chosen preorders for the sender
+		///
+		/// Parameters:
+		/// `origin`: Sender
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn claim_chosen_preorders(origin: OriginFor<T>) -> DispatchResult {
+			ensure!(
+				!CanPreorderOriginOfShells::<T>::get(),
+				Error::<T>::PreorderOriginOfShellNotAvailable
+			);
+			let sender = ensure_signed(origin)?;
+			let overlord = Self::get_overlord_account()?;
+
+			// Has Spirit Collection been set
+			let spirit_collection_id = Self::get_spirit_collection_id()?;
+			// Must have a spirit to purchase a Origin of Shell
+			ensure!(
+				pallet_uniques::Pallet::<T>::owned_in_class(&spirit_collection_id, &sender).count() >
+					0,
+				Error::<T>::MustOwnSpiritToPurchase
+			);
+			// Must have origin of shell collection created
+			let origin_of_shell_collection_id = Self::get_origin_of_shell_collection_id()?;
+			ensure!(
+				(pallet_uniques::Pallet::<T>::owned_in_class(
+					&origin_of_shell_collection_id,
+					&sender
+				)
+				.count() == 0) || LastDayOfSale::<T>::get(),
+				Error::<T>::OriginOfShellAlreadyPurchased
+			);
+			// Get price of hero origin of shell
+			let hero_origin_of_shell_price = T::HeroOriginOfShellPrice::get();
+			// Check if any preorders were chosen
+			let preorders_iter = PreorderResults::<T>::drain_prefix(sender.clone());
+			for (preorder_id, preorder) in preorders_iter {
+				let preorder_status = preorder.clone().preorder_status;
+				match preorder_status {
+					PreorderStatus::Chosen => {
+						// Next NFT ID of Collection
+						let nft_id =
+							pallet_rmrk_core::NextNftId::<T>::get(origin_of_shell_collection_id);
+						// Get payment from owner's reserve
+						<T as pallet::Config>::Currency::unreserve(
+							&sender,
+							hero_origin_of_shell_price,
+						);
+						<T as pallet::Config>::Currency::transfer(
+							&sender,
+							&overlord,
+							hero_origin_of_shell_price,
+							ExistenceRequirement::KeepAlive,
+						)?;
+						// Mint Origin of Shell and transfer Origin of Shell to new owner
+						pallet_rmrk_core::Pallet::<T>::mint_nft(
+							Origin::<T>::Signed(overlord.clone()).into(),
+							preorder.owner.clone(),
+							origin_of_shell_collection_id,
+							None,
+							None,
+							preorder.metadata,
+						)?;
+
+						Self::deposit_event(Event::OriginOfShellMinted {
+							collection_id: origin_of_shell_collection_id,
+							nft_id,
+							owner: preorder.owner,
+						});
+					},
+					PreorderStatus::NotChosen => {
```

Comment:
If the not-chosen preorders are not processed in this function, is it better to just separate the two kinds of preorders to two storage map?

Reply:
Makes sense. I can create 2 separate storages then.